src/snowflake/snowpark/_internal/debug_utils.py

```diff
@@ -198,3 +202,71 @@ def get_df_transform_trace_message(
             "environment variable."
         )
     return "\n".join(debug_info_lines)
+
+
+def _format_source_location(src: Optional[proto.SrcPosition]) -> str:
+    """Helper function to format source location information."""
+    if src is None:
+        return ""
+
+    from snowflake.snowpark._internal.ast.utils import __STRING_INTERNING_MAP__
+
+    filename_map = {v: k for k, v in __STRING_INTERNING_MAP__.items()}
+    # if we cannot find the file, we use a placeholder
+    filename = filename_map.get(src.file, "UNKNOWN_FILE")
+    lines_info = f"{filename}: line {src.start_line}"
+    if src.end_line > src.start_line:
+        lines_info = f"{filename}: lines {src.start_line}-{src.end_line}"
+    return lines_info
+
+
+def get_python_source_from_sql_error(top_plan: "SnowflakePlan", error_msg: str) -> str:
+    """
+    Extract SQL error line number and map it back to Python source code. We use the
+    helper function get_plan_from_line_numbers to get the plan from the line number
+    found in the SQL compilation error message. We then extract the source lines
+    and columns using the ast_id associated with the plan. We return a message with
+    the affected Python lines if found, otherwise an empty string.
+
+    Args:
+        plan: The top level SnowflakePlan object that contains the SQL compilation error.
+        error_msg: The error message from the SQL compilation error.
+
+    Returns:
+        Error message with the affected Python lines numbers if found, otherwise an empty string.
+    """
+    sql_compilation_error_regex = re.compile(
+        r""".*SQL compilation error:\s*error line (\d+) at position (\d+).*""",
+    )
+    match = sql_compilation_error_regex.match(error_msg)
+    if not match:
+        return ""
+
+    sql_line_number = int(match.group(1)) - 1
+
+    try:
+        from snowflake.snowpark._internal.utils import (
+            get_plan_from_line_numbers,
+        )
```

Comment:
remove this try catch since we will wrap this in try-except inside snowflake plan